### ISSUE TITLE
Issue using `OneHotEncoder` with `min_frequency` specified

### ISSUE THREAD

**Issue by aohma** [onnx/sklearn-onnx#1026]
When using `OneHotEncoding` with the `min_frequency` keyword specified, the prediction is different (wrong) after converting a pipeline to onnx. Here is an example to reproduce the behavior:

```
import numpy as np
from onnxruntime import InferenceSession
from pandas import DataFrame
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from skl2onnx import to_onnx
from skl2onnx.algebra.type_helper import guess_initial_types

data = DataFrame(
    [
        dict(CAT1="a", CAT2="a", num1=0.5, num2=0.6, y=0),
        dict(CAT1="b", CAT2="b", num1=0.4, num2=0.8, y=1),
        dict(CAT1="a", CAT2="c", num1=0.5, num2=0.56, y=0),
        dict(CAT1="a", CAT2="d", num1=0.55, num2=0.56, y=1),
        dict(CAT1="a", CAT2="d", num1=0.35, num2=0.86, y=0),
        dict(CAT1="a", CAT2="d", num1=0.5, num2=0.68, y=1),
    ]
)

cat_cols = ["CAT1", "CAT2"]
train_data = data.drop("y", axis=1)


categorical_transformer = Pipeline(
    [("onehot", OneHotEncoder(min_frequency=2, sparse=False, handle_unknown="ignore"))]
)
preprocessor = ColumnTransformer(
    transformers=[("cat", categorical_transformer, cat_cols)], remainder="passthrough"
)
pipe = Pipeline([("preprocess", preprocessor), ("rf", RandomForestClassifier())])
pipe.fit(train_data, data["y"])

for c in train_data.columns:
    if c not in cat_cols:
        train_data[c] = train_data[c].astype(np.float32)
init = guess_initial_types(train_data,None)

onx2 = to_onnx(
    pipe, initial_types=init, options={RandomForestClassifier: {"zipmap": False}}
)

inputs = {c: train_data[c].values.reshape((-1, 1)) for c in train_data.columns}

sess2 = InferenceSession(onx2.SerializeToString(), providers=["CPUExecutionProvider"])

got2 = sess2.run(None, inputs)

print(pipe.predict(train_data))
print(got2[0])

identical = (np.array(pipe.predict(train_data)) == np.array(got2[0])).all()
print('Identical: ', identical)
```

My guess is that collecting infrequent categories is not implemented in onnx/skl2onnx, causing the onnx model to encode all categories including the infrequent ones. The same error occurs when `max_categories` is used.

Could this be fixed, making the onnx model behave as the input pipeline?
It would also be good if a warning/error is raised when using these keywords until a fix is available, instead of silently producing an inconsistent model.

**Comment by perhba** [comment]
I think the bug lies in [convert_sklearn_one_hot_encoder](https://github.com/onnx/sklearn-onnx/blob/262abfc78b0079d1f2eb8113bfbf67abd50b1bd3/skl2onnx/operator_converters/one_hot_encoder.py#L19). `one_hot_encoder` derives the list of categories from `categories_` attribute to `sklearn.preprocessing.OneHotEncoder`, without taking into account the `infrequent_categories_` attribute.  `infrequent_categories_` should be mapped  to a unknow-category or ignored. 


````python
cat = pd.DataFrame( data=np.random.choice(["A","B","C","D"], size=100_000, p=[0.01, 0.01, 0.49, 0.49]), columns=['cat'])
encoder = OneHotEncoder(
    handle_unknown='error', # ignore, error, infrequent_if_exist
    min_frequency=0.2,
    sparse_output=False)

transformed_cat = encoder.fit_transform(cat)
transformed_cat

print(encoder.categories_)
print(encoder.infrequent_categories_)
# OUT:
# [array(['A', 'B', 'C', 'D'], dtype=object)]
# [array(['A', 'B'], dtype=object)]

**Comment by cgrabows-ml** [comment]
This issue appears to occur anytime categories are present in inference data that were not available during training.


import numpy as np
import onnxruntime as ort  # type: ignore
import pandas as pd
from skl2onnx.common.data_types import (  # type: ignore
    StringTensorType,
    Int64TensorType
)
from skl2onnx import convert_sklearn  # type: ignore

from sklearn.compose import ColumnTransformer  # type: ignore
from sklearn.linear_model import LogisticRegression  # type: ignore
from sklearn.pipeline import Pipeline  # type: ignore
from sklearn.preprocessing import OneHotEncoder  # type: ignore
from skl2onnx import convert_sklearn  # type: ignore

# Generate some sample data
data = {
    "num": [2,    8,   4,   5,   7,   5,   5,   6,   4,   8],
    "cat": ["A", "B", "A", "B", "A", "A", "B", "X", "Y", "Z"],  # String features
    "target": [0, 1,   0,   1,   0,   0,   1,   1,   0,   0],
}

df = pd.DataFrame(data)
X = df[["cat", "num"]]
y = df["target"]
categorical_features = ["cat"]
categorical_transformer = Pipeline(
    steps=[("onehot", OneHotEncoder(handle_unknown="ignore"))]
)
transformers = [("cat", categorical_transformer, categorical_features)]
preprocessor = ColumnTransformer(transformers=transformers)
model = LogisticRegression()
pipeline = Pipeline(steps=[("preprocessor", preprocessor), ("classifier", model)])

i = 7
X_train, X_test = X[:i], X[i:]
y_train, y_test = y[:i], y[i:]
pipeline.fit(X_train, y_train)
initial_types = [("cat", StringTensorType([None, 1])),
                 ("num", Int64TensorType([None, 1]))]
onx = convert_sklearn(pipeline, initial_types=initial_types).SerializeToString()

sess = ort.InferenceSession(onx)
input = {
    "cat": X_test["cat"].values.astype(dtype=np.str_).reshape(len(X_test), 1),
    "num": X_test["num"].values.astype(dtype=np.int64).reshape(len(X_test), 1),
}
result = sess.run(None, input)
pipeline_predictions = pipeline.predict_proba(X_train)[:,1]
predictions = [r[1] for r in result[1]]
# Print the predictions
print(predictions)
print(pipeline_predictions)
print("looking for errors...")
for (i, p) in enumerate(predictions):
    if round(predictions[i], 5) != round(pipeline_predictions[i], 5):
        print(i, predictions[i], pipeline_predictions[i])

**Comment by ellese-c** [comment]
I've also been getting an error when using `one_hot_encoder` with `max_categories` converted to onnx. It throws an error when a category is encountered at inference time that wasn't in training. Has anyone found a work around/fix?

**Comment by aohma** [comment]
@cgrabows-ml You are using test data in the onx model and train data in the pipeline model. When I change to `pipeline_predictions = pipeline.predict_proba(X_test)[:,1]` the code finds no inconsistencies.

@ellese-c I have not experienced this. Are you specifying `handle_unknown='ignore`? The default is 'error', which should throw errors for categories not present during training.

**Comment by aohma** [comment]
A possible workaround for this issue seems to be manually selecting frequent categories and setting handle_unknown='ignore'. Infrequent categories are then encoded as all zeros. Here is a minimum example:

```
import numpy as np
from onnxruntime import InferenceSession
from pandas import DataFrame
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.ensemble import RandomForestClassifier
from skl2onnx import to_onnx
from skl2onnx.algebra.type_helper import guess_initial_types

data = DataFrame(
    [
        dict(CAT1="a", CAT2="a", num1=0.5, num2=0.6, y=0),
        dict(CAT1="b", CAT2="b", num1=0.4, num2=0.8, y=1),
        dict(CAT1="a", CAT2="c", num1=0.5, num2=0.56, y=0),
        dict(CAT1="a", CAT2="d", num1=0.55, num2=0.56, y=1),
        dict(CAT1="a", CAT2="d", num1=0.35, num2=0.86, y=0),
        dict(CAT1="a", CAT2="d", num1=0.5, num2=0.68, y=1),
    ]
)

cat_cols = ["CAT1", "CAT2"]
train_data = data.drop("y", axis=1)

# Find frequent categories
min_frequency=2
categories = []
for k in cat_cols:
    categories.append(train_data[k].value_counts()[train_data[k].value_counts()>=min_frequency].index.tolist())

categorical_transformer = Pipeline(
    [("onehot", OneHotEncoder(categories=categories,handle_unknown='ignore', sparse_output=False))]
)
preprocessor = ColumnTransformer(
    transformers=[("cat", categorical_transformer, cat_cols)], remainder="passthrough"
)
pipe = Pipeline([("preprocess", preprocessor), ("rf", RandomForestClassifier())])
pipe.fit(train_data, data["y"])

init = guess_initial_types(train_data,None)

onx = to_onnx(
    pipe, initial_types=init, options={RandomForestClassifier: {"zipmap": False}}
)

sess = InferenceSession(onx.SerializeToString(), providers=["CPUExecutionProvider"])

# Check consistency with train data
inputs = {c: train_data[c].values.reshape((-1, 1)) for c in train_data.columns}
got = sess.run(None, inputs)
print(pipe.predict(train_data))
print(got[0])

# Check consistency with inference data including previously unknown categories
inference_data = DataFrame(
    [
        dict(CAT1="b", CAT2="a", num1=0.2, num2=0.6),
        dict(CAT1="b", CAT2="not_in_train", num1=0.4, num2=0.8),
        dict(CAT1="a", CAT2="d", num1=0.75, num2=0.8),
        dict(CAT1="not_in_train", CAT2="d", num1=0.5, num2=0.6),
        dict(CAT1="a", CAT2="c", num1=0.5, num2=0.8),
        dict(CAT1="a", CAT2="d", num1=0.4, num2=0.56),
    ]
)

inputs = {c: inference_data[c].values.reshape((-1, 1)) for c in inference_data.columns}
got = sess.run(None, inputs)
print(pipe.predict(inference_data))
print(got[0])
```

which prints
```
[0 1 0 1 0 1]
[0 1 0 1 0 1]
[0 1 1 1 0 1]
[0 1 1 1 0 1]
```
It should be straight forward to also select the top k most frequent categories to mimic `max_categories`

**Comment by ellese-c** [comment]
@aohma The error occurs when using `OneHotEncoder` with `handle_unknown='infrequent_if_exist'`. As a minimal example, 

```from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import FloatTensorType, StringTensorType
from sklearn import preprocessing
import onnxruntime as rt
import numpy as np

one_hot = preprocessing.OneHotEncoder(sparse=False, handle_unknown='infrequent_if_exist', max_categories=3)
one_hot.fit(np.array(['a', 'a', 'b', 'b', 'c']).reshape(-1, 1))
initial_type = [('catfeat', StringTensorType(shape=[None, 1]))]

model_onnx = convert_sklearn(one_hot, initial_types=initial_type)
sess = rt.InferenceSession(model_onnx.SerializeToString(), providers=["CPUExecutionProvider"])
input_name = sess.get_inputs()[0].name
label_name = sess.get_outputs()[0].name
pred_onx = sess.run([label_name], {input_name: np.array(['a', 'a', 'b', 'b', 'd']).reshape(-1, 1)})[0]```

produces the error
```Fail: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running OneHotEncoder node. Name:'OneHotEncoder' Status Message: Unknown Category and zeros = 0.```.

The standard `OneHotEncoder` handles this case by encoding all unknown categories as all zeros. The workaround you mentioned does work, but isn't exactly the same behaviour as using 'infrequent_if_exist'.

**Comment by aohma** [comment]
You are completely right @ellese-c , the above code fails for me as well. I guess `handle_unknown='infrequent_if_exist'` has the same problems as `min_frequency` and `max_categories` in that infrequent categories are not considered, although your example fails while mine produces inconsistent predictions without failing. There is thus a general problem when trying to map to an infrequent column. In addition to the issues described by @perhba above, I guess changes are also needed at runtime to correctly handle infrequent and unknown categories.

I am aware that the proposed workaround is not identical to the behavior of the sklearn `OneHotEncoder` with `handle_unknown='infrequent_if_exist'` or when using `min_frequency`/`max_categories`, since the workaround simply sets infrequent and unknown entries to all zeros while the encoder maps them to an infrequent feature column. However, I think the model gets the same information, similar to using `drop=...`. This is adequate for my application, but I guess there could be situations where such a difference does matter and the onnx encoder should ideally be exactly equivalent to the sklearn encoder.
